### PR TITLE
Add test code for grammar specification

### PR DESCRIPTION
**Pull request by sigmaith**
**What this PR does / why we need it:**

Here's what we need to do for schema validation.

**Device A. Devices that validate user-defined schemas**
**Device B. Devices that validate user-altered values with pre-defined schema**

Regarding **device A**, we would like to support the following specifications.

### Element level schema definition:
- A type definition syntax similar to TypeScript is used.
- Yorkie’s primitive types can be used.
- Yorkie's element-level default types can be utilized.
   (yorkie.Array, yorkie.Object, yorkie.Counter, yorkie.Text, yorkie.Tree)
- Additionally, user-defined types are allowed.
- Undefined types are not permitted.

### Tree level schema definition:
- Further discussion is required.

In detail,  I have written test code to validate the following grammar specifications.

### CheckList
**TypeScript-Like Grammar**
- [x] Add TypeScript and other type tests to ensure errors are correctly thrown.
 - [x] `;` (semicolon supported)
 - [x] `Array<Todo>, Array<Array<Todo>>` support 
     (as -is: only`Todo[]`supported)
 - [x] `?` optional properties 
 - [x] `|` union operator 
 - [x] Literal type 
 - [x] annotation (`#`, `//` skip)
 - [ ] `&` Intersection Types
 - [ ]  [Utility Types](https://www.typescriptlang.org/docs/handbook/utility-types.html#handbook-content)

**Yorkie Primitive Types**
- [x] string, number, boolean, null, bigint, Uint8Array, Date
- [x] An error occurs when unsupported types (long, bytes) are set.

**Yorkie Element &  Generic Parameter**
  - [x] `yorkie.Object`, `yorkie.Array`, `yorkie.Counter`, `yorkie.Text`, `yorkie.Tree`
  - [x] `yorkie.Array` -> fail, `yorkie.Array<Todo>` -> pass
  - [x] `yorkie.Object`-> fail,`yorkie.Object<Todo>` -> pass
  - [x] `yorkie.Counter` -> pass, `yorkie.Counter<Todo>` -> fail 
  - [x] `yorkie.Text<Attribute>` -> pass (Attribute: pre-defined type)
  - [x] `yorkie.Text<{bold: boolean; italic: boolean;}>` -> pass
     

**User-Defined Type**
- [x] user-defined type
   e.g. type Todo { title: string; completed: boolean; }
- [x] undefined type error

**Error Handling**
- [x] Use reserved word
- [x] Type cycle

### Follow-up Implementations Needed
- Refactor grammar to pass all tests
- Customize ANTLR's baseErrorListener to define and throw **parseError(syntaxError)**
- Implement an AST visitor or listener to validate **semanticError**(e.g. use undefined type, type cycle)
- Develop an **ErrorReporter** to comprehensively report all errors

## Summary by CodeRabbit

- **New Features**
	- Introduced a comprehensive overhaul of the schema grammar, enhancing parsing capabilities with new types and structures.
	- Added support for Yorkie primitive types and user-defined types in the schema validation.
	- Expanded the test suite for schema validation, covering various scenarios and edge cases.

- **Bug Fixes**
	- Improved error handling in the validation process to provide more descriptive error messages.

- **Documentation**
	- Enhanced clarity in the test structure and coverage for schema validation scenarios.

### CODE REVIEW

File: antlr/Schema.g4 [hunk 1]
```diff
@@ -1,46 +1,110 @@
...
+	| yorkieTreeType;
+
+yorkieObjectType:
+	YORKIE_OBJECT LT (typeReference | objectLiteralType) GT;
+yorkieArrayType:
+	YORKIE_ARRAY LT (typeReference | objectLiteralType) GT;
+yorkieCounterType: YORKIE_COUNTER;
+yorkieTextType:
+	YORKIE_TEXT LT (typeReference | objectLiteralType) GT;
+yorkieTreeType: YORKIE_TREE LT GT;
```
Comment (hackerwins) [inline antlr/Schema.g4]: It would be better to insert a newline at the end of files.
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

Comment (sigmaith) [reply antlr/Schema.g4]: When I checked in IDE, there was the last blank line in validator.ts and schema.g4, so I only added it to the test code. Thank you.

**Comment by hackerwins** [comment]
This was merged by https://github.com/yorkie-team/yorkie-schema/pull/4.
